### ISSUE TITLE
[I18n] Add i18n support for Messages

### ISSUE THREAD

**Issue by UkonnRa** [ory/kratos#691]
**Is your feature request related to a problem? Please describe.**

Now Messages are something like:

```json
[
  {
    "id": 4000006,
    "text": "The provided credentials are invalid, check for spelling mistakes in your password or username, email address, or phone number.",
    "type": "error",
    "context": {}
  }
]
```

So if `text` contains some variables, it will be impossable to do the i18n work

**Describe the solution you'd like**

Use [ICU format](http://site.icu-project.org/) to solve the problem

```json5
[
  {
    "id": 4000006, // This is a 1/1 mapping to `i18nText`, in another word, one ID can map to one and only one `i18nText`
    "text": "Hello world, Foo Bar", // A meanful description for developer, it should NOT be used as i18n directly
    "i18nText": "Hello, {firstName} {lastName}",
    "i18nData": {
        "firstName": "Foo",
        "lastName": "Bar"
    },
    "type": "error",
    "context": {}
  }
]
```

**Describe alternatives you've considered**

No

**Additional context**

React i18n-next: https://react.i18next.com/misc/using-with-icu-format

**Comment by zepatrik** [comment]
Actually the parameters `firstName` and `lastName` would be under the `"context"` key. But I agree with you that we should use a standard instead of our own solution, especially when it does exactly the same thing. The refactoring would not be much work, just enforcing the standard at this type: https://github.com/ory/kratos/blob/master/text/message.go#L35-L40
Would you be up for a PR @UkonnRa ?

**Comment by UkonnRa** [comment]
I wonder what is the current i18n solution right now? How Kratos handles variables, plural, gender right now? Or right now all Messages are static strings and have no variables at all?

**Comment by zepatrik** [comment]
Right now we just have IDs so you can map messages to your own strings. You have to interpret the context on your own and we don't have a neat overview over all the IDs and associated contexts. The message you see is just a sensible default you can display if you don't want to bother rewriting all the messages. Until now, we had no complex messages that would require plural or gender forms, but we will get to that point eventually. Everything happens in the [text](https://github.com/ory/kratos/tree/master/text) package, including all the constants.

**Comment by aeneasr** [comment]
Yeah we actually don't want to deal with complex i18n scenarios. Most messages are errors ("not an email") or success messages ("update successful") and not stuff like welcome messages or others. In fact, i18n is a non-goal for kratos but we obviously want to make it easy for developers to translate messages. I think the simpler we keep stuff the easier it is to translate! But I agree that a standardized format would be nice!

**Comment by UkonnRa** [comment]
Let me give a simple example. In [message_login](https://github.com/ory/kratos/blob/4f56651fff130b114de27e39dd1cc6df725bf556/text/message_login.go#L20), we have a message `The login request expired %.2f minutes ago, please try again.`. Because containing a variable `ago.Minutes()`, it is impossible to use standard i18n tools. So using some standard format is a must, I think.

**Comment by krhubert** [comment]
@aeneasr What if Kratos could start with just listing of all pairs id/message that API endpoints can return on error? I think, it would make translation way easier for others without implementing i18n directly into kratos.

**Comment by aeneasr** [comment]
You can currently find those in the message_* files: https://github.com/ory/kratos/tree/master/text

**Comment by krhubert** [comment]
I've checked the code and I saw these files, but not all messages are there. For example:

```
4000005 The password can not be used because the password has been found in data breaches and must no longer be used.
```

Part of this one is located in [validator.go](https://github.com/ory/kratos/blob/master/selfservice/strategy/password/validator.go#L188).

So A lot can be found in the package you gave, but not all of them.

**Comment by aeneasr** [comment]
Ah, nice find - for all user-related error messages (so not system errors) I think we could move all to text/c :) Some of those messages have been added before we added text/

**Comment by koudshoorn** [comment]
Hi, sorry for tuning in on this closed issue. In its current form, translating the reason for a validation error is indeed very difficult. For context values like names or amounts it is not a problem to directly plug them into the translation. However, a validation error reason does require additional translation.

Would it be a solution to provide the `reason` or perhaps other contexts as an additional message node? That way a developer can fully translate the validation message (assuming they know they can expect a nested message). 

e.g.
```json
{
    "id": 4000005,
    "text": "The password can not be used because {reason}",
    "type": "error",
    "context": {
        "reason": {
              "id": 49999999
              "text": "the password has been found in data breaches and must no longer be used."
              "type": "error",
              "context": {}
         }
    }
}
```

**Comment by groue** [comment]
I also face a localization difficulty. Here are two error messages from the registration flow that share the same id:

```
{
  "id": 4000001,
  "text": "\"test\" is not valid \"email\"",
  "type": "error"
}

{
  "id": 4000001,
  "text": "length must be >= 1, but got 0",
  "type": "error"
}
```

Both messages share the same 4000001 id, which is [`ErrorValidationGeneric `](https://github.com/ory/kratos/blob/acf92618559f59bc51ac28bf8251a4a1437f44cf/text/id.go#L95).

How can I localize those error messages?

I don't want to state obvious truths, but error localization is so very important. How can a user proceed if they don't understand the user interface? We're not here to lock users out, right?

**Comment by McKean** [comment]
I'm with @aeneasr that this should not be solved by kratos but should be handled by the ui or server side application exclusively.

However, doing so would challenge kratos to not really speak a language but provide something along the lines of this:
```
{
  "id": 4000006,
  "key": "invalid-credentials",
  "type": "error",
  "context": {
    "someVar": "bar"
  }
}
```
text could still be included but eventually should be deprecated that way the stance is clearly taken.

**Comment by McKean** [comment]
If this makes sense I'd be happy to take a stab at it in the context of a new feature request :shrug:

**Comment by woylie** [comment]
It would at least be great if it was possible to translate all messages with the information Kratos returns without resorting to Regex.

**Comment by McKean** [comment]
@woylie with keys in place and not using any libraries client side such as i18next you could still just do case switching to check for the key and then use your string and replace predefined placeholders in your string handlebars style.

eg.
1. check for `invalid-credentials`
2. use your string "those creds aren't valid bro, have a drink at the {{someVar}}"
3. replace {{someVar}} with someVar (bar).

I would still recommend some i18n library, but your concern should be addressed.

**Comment by woylie** [comment]
I would still prefer a solution where I have a usable example from which I can derive translations (I might not even need other languages than English in some applications). But I would also like to be able to use the [published messages in JSON format](https://www.ory.sh/docs/kratos/concepts/ui-user-interface#machine-readable-format) to generate the code necessary to turn those messages directly into whatever i18n format I need. At the moment, this is impossible, because the messages don't consequently use variables, and even if they do, they don't always do it in a usable way.

For example, there are several messages with expiration date times, but no variables are in used in the message string:

```json
{
    "id": 4010001,
    "text": "The login flow expired 1.02 minutes ago, please try again.",
    "type": "error",
    "context": {
      "expired_at": "2020-01-01T00:59:59Z"
    }
  }
```

And this is completely useless:

```json
{
    "id": 4000005,
    "text": "The password can not be used because {reason}.",
    "type": "error",
    "context": {
      "reason": "{reason}"
    }
  }
```

The way the messages are returned at the moment, I can only generate source strings from the documentation for part of the messages, while I need to explicitly match on the message IDs and hard code the source strings for the odd ones. Which by extension means that any time I upgrade Kratos, I need to manually check whether there are any new messages that might need special treatment, instead of just copying the new published JSON and letting my generator do the rest.

So in summary, the current message format would work fine for me, but the handling of variables needs to be consistent throughout all messages.

**Comment by McKean** [comment]
yeah so adding a `"key"` should do that for ya... or am I missing something?

**Comment by woylie** [comment]
Well, in your proposal, you _have_ to add translations. No usable defaults. The user would just see "invalid-credentials", unless you explicitly translate that key into a user-readable message.

---

For context, and detached from your solution, in my application, the base language for all texts in the UI is English. Those strings will automatically be extracted from the source code and saved in gettext `.pot` files. We don't have to add English translations, because English is already the source.

For Kratos messages, the application reads the json published in the documentation at compile time, takes the message text and converts it into the gettext format. For messages without the need for variables, this works just fine. This:

```json
{
    "id": 12345678,
    "text": "The provided authentication code is invalid, please try again.",
    "type": "error",
}
```

Will be converted to this `.pot` content:

```
msgid "The provided authentication code is invalid, please try again."
msgstr ""
```

In the English `.po` file, I can just leave it as it is. In the Japanese `.po` file, I just add the Japanese version as `msgstr`:

```
msgid "The provided authentication code is invalid, please try again."
msgstr "入力された認証コードが無効です。もう一度実行してください。"
```

If variables are properly used, I can work with that, too. This:

```json
{
    "id": 1050018,
    "text": "Remove security key \"{name}\"",
    "type": "info",
    "context": {
      "name": "{name}"
    }
 },
```

Will be converted to:

```
msgid "Remove security key \"%{name}\""
msgstr ""
```

With that, I would only need a single function that takes the raw Kratos message returned by the API, reads the text and context fields, and passes these values to the gettext function to get a translated message.

But with hardcoded values in messages templates like this:

```json
{
    "id": 4010001,
    "text": "The login flow expired 1.02 minutes ago, please try again.",
    "type": "error",
    "context": {
      "expired_at": "2020-01-01T00:59:59Z"
    }
  }
```

I cannot do anything in an automated way. This would just result in:

```
msgid "The login flow expired 1.02 minutes ago, please try again.""
msgstr ""
```

---

Anyway, I digress. Of course, this would all work with just a key as msgid. That still burdens the user with coming up with their own English texts, though. Well, 🤷 , I'm tired and probably don't make any sense anymore.

**Comment by McKean** [comment]
I would suggest not diving into gettext and po jadi jada jada... the client can deal with this as it wants.
In the example I provided we can still keep a default text (this is what I mentioned in my previous message)
> text could still be included but eventually should be deprecated that way the stance is clearly taken.
kratos should just not deal with translations at all, the client is way more capable to deal with that.

**Comment by zepatrik** [comment]
If I understand correctly, it would be enough to include the raw format string, e.g. https://github.com/ory/kratos/blob/74ae85236928f55156284dd609ca88cafe90b360/text/message_login.go#L115 in the autogenerated document? I guess that can be done quite easily.

**Comment by woylie** [comment]
> I would suggest not diving into gettext and po jadi jada jada... the client can deal with this as it wants.

I was describing how the Kratos client application I'm working on is translating Kratos messages using Gettext, not proposing to add Gettext to Kratos.

> If I understand correctly, it would be enough to include the raw format string, e.g.
> 
> https://github.com/ory/kratos/blob/74ae85236928f55156284dd609ca88cafe90b360/text/message_login.go#L115
> in the autogenerated document? I guess that can be done quite easily.

That would be one step better. However, as far as I remember, there are also messages with more than one variable, and there are others where one variable is used in the format string, but multiple fields are set in the context. It would be better if the symbols would be substituted with the variable names as used in the context map. In the example you linked, the minute value isn't even in the context map, which would lead to weird results if you try to automate this on the client side as I described. To be useful, the output would have to be:

```json
  {
    "id": 4010001,
    "text": "The login flow expired {minutes} minutes ago, please try again.",
    "type": "error",
    "context": {
      "expired_at": "2020-01-01T00:59:59Z",
      "minutes": 2
    }
  }
```

But then the question is how to deal with singular/plural.

**Comment by zepatrik** [comment]
I see. We generate the file at this point: https://github.com/ory/kratos/blob/1a5ead43a60e7a0388617877a9f16d1dec61459b/cmd/clidoc/main.go#L165
If you have any ideas how to implement your suggestion, feel free to open a (draft) PR. I do not see a trivial way though, especially to name fields.

**Comment by woylie** [comment]
Would it make sense to replace sprintf formatting used in [messages like this](https://github.com/ory/kratos/blob/1a5ead43a60e7a0388617877a9f16d1dec61459b/text/message_registration.go#L20) with [text templates](https://pkg.go.dev/text/template)? We could then write the raw template string to `messages.json`.

Alternatively, I could see value in replacing the current Message implementation including the message writer with something like [go-i18n](https://github.com/nicksnyder/go-i18n). That would immediately take care of the plural as well. All messages can be extracted to a JSON or TOML file with a CLI tool. The goal here would not be that Kratos returns translated messages, it can always return the default message as before. The goal is to have an automated output of all messages in a format that can be used for code generation.

**Comment by McKean** [comment]
I really believe we should not try to solve messaging within kratos. We can keep things simple add keys and be done with it. All customization should then just be handled client side.
I see the value in having the capability to customize the defaults, and perhaps exposing this in one of the ways suggested would be a nice addition.

Would we agree on this:
* If you want to deal with plurals or introduce additional languages, you need to do so outside of kratos.

Seems pretty fair?

Just trying to get on the same page.

**Comment by woylie** [comment]
I agree that Kratos doesn't need to translate messages to other language, but I still think there should be decent default messages, so that you can get up running quickly without having to first write out all messages yourself. And that would potentially include plurals, if counts are used in a default message.

**Comment by McKean** [comment]
perhaps introducing message templates is a sound solution and in the config you can choose which template to use.
That would allow community efforts to establish multiple templates eg. `messages-en.json` `messages-fr.json` messages-keys.json` - everybody gets what the want :smile: 

Kratos would be language agnostic. Plurals is just oping a whole new world where things may differ based on languages. I think we should refrain from introducing additional packages as this would just widen the threat surface. The ory ecosystem is proud to be relatively third party independent, for good reason.

**Comment by woylie** [comment]
Sure. I don't think there are any counts in the messages anyway, with the exception of the `expired 1.07 minutes ago` messages, which in my opinion should just default to `The flow has expired.`.

**Comment by renom** [comment]
+1 for i18n support and adding Russian translation
